File: frontend/src/services/stream.js
// Live push client for WS /api/v1/stream (backend/app/routers/stream.py).
// Ingestion publishes a tick/alert to Redis the moment it happens; this
// relays straight through the backend to here -- what lets Dashboard.jsx
// react to real trade activity instead of polling on a flat timer.

function streamUrl() {
    // Same host the page itself was loaded from, upgraded to ws:/wss: to
    // match the page's own http:/https: -- works unchanged through
    // nginx's /api/ proxy (frontend/nginx.conf) or an ingress routing
    // /api straight to the backend service, since it's a relative path
    // either way, not a hardcoded host.
    const protocol = window.location.protocol === "https:" ? "wss:" : "ws:";
    return `${protocol}//${window.location.host}/api/v1/stream`;
}

// Opens the connection and keeps it alive; calls onEvent(payload) for
// every real tick/alert message (the server's own keepalive pings are
// swallowed here, never passed through). Returns an unsubscribe function
// that closes the connection and stops reconnecting.
export function subscribeToEvents(onEvent) {
    let socket = null;
    let reconnectTimer = null;
    let stopped = false;

    function connect() {
        const token = localStorage.getItem("token");
        if (!token) {
            // Not signed in -- nothing to authenticate the socket with.
            // AuthContext's own sign-in flow will trigger a fresh
            // Dashboard mount (and so a fresh connect()) once it is.
            return;
        }

        socket = new WebSocket(streamUrl());

        socket.onopen = () => {
            // First frame is the auth handshake the server's
            // _authenticate() is waiting for -- browsers can't set an
            // Authorization header on a WS handshake, so the token goes
            // here instead.
            socket.send(JSON.stringify({ token }));
        };

        socket.onmessage = (event) => {
            let payload;
            try {
                payload = JSON.parse(event.data);
            } catch {
                return;
            }

            if (payload.type === "ping") return;

            onEvent(payload);
        };

        socket.onclose = (event) => {
            if (stopped) return;

            if (event.code === 4401) {
                // Server rejected the token -- missing, expired, or
                // revoked (see backend/app/routers/stream.py's
                // _authenticate/REAUTH_INTERVAL_SECONDS). Retrying with
                // the exact same stored token would just get rejected
                // again every time, forever -- reconnecting here would
                // mean hammering the backend with a fresh handshake + DB
                // lookup every 3s indefinitely. Stop until a fresh
                // sign-in replaces the token; AuthContext's sign-in flow
                // re-mounts Dashboard (and so calls connect() fresh)
                // once that happens.
                return;
            }

            // Reconnect on any other close -- a network blip or a
            // backend pod restart shouldn't permanently strand the
            // dashboard on "no live updates" until a manual page refresh.
            reconnectTimer = setTimeout(connect, 3000);
        };

        socket.onerror = () => {
            socket.close();
        };
    }

    connect();

    return function unsubscribe() {
        stopped = true;

        if (reconnectTimer) clearTimeout(reconnectTimer);
        if (socket) socket.close();
    };
}

File: backend/app/routers/stream.py
"""
Live push relay: WS /api/v1/stream.

ingestion (finnhub_ingestion.py) publishes each tick/alert to Redis right
after writing it to Postgres. This endpoint is the other end of that:
one long-lived WebSocket connection per browser tab, subscribed to the
same Redis channels, relaying each message down to the client the moment
it arrives -- what lets the dashboard drop its old 5s poll in favor of
updating only when something actually happened.

Auth doesn't go through the usual Depends(get_current_user) path -- a
browser's native WebSocket API can't set an Authorization header on the
handshake request, unlike a normal fetch/XHR call. Instead: the
connection is accepted first, then the client's first text frame must be
{"token": "<jwt>"}, validated with the exact same decode + user lookup
every REST endpoint uses (get_user_from_token). Anything else --
timeout, malformed frame, bad/expired token -- closes the connection
with a 4401 (app-defined WS close code in the 4000-4999 range reserved
for that) rather than ever subscribing to Redis for an unauthenticated
caller. Unlike a REST call, though, a WS connection can sit open for
hours -- so that same check also re-runs every REAUTH_INTERVAL_SECONDS
for the life of the connection, not just once at handshake (see
_validate_token/REAUTH_INTERVAL_SECONDS below).
"""

import asyncio
import json
import time

from fastapi import APIRouter, WebSocket, WebSocketDisconnect
from redis import RedisError
from starlette.concurrency import run_in_threadpool

from app.core.redis_client import get_async_redis
from app.core.security import get_user_from_token
from app.database.session import SessionLocal

router = APIRouter(tags=["Stream"])

CHANNELS = ["ticks", "alerts"]
AUTH_TIMEOUT_SECONDS = 10

# How long to wait for a Redis message before sending a keepalive ping
# instead. Below nginx's default proxy idle timeouts (and the ingress's)
# so a quiet market (nothing traded in a while) never looks like a dead
# connection to anything proxying this.
KEEPALIVE_SECONDS = 15

# Re-checked on this cadence for the life of the connection (see the main
# loop below) -- without this, a JWT valid at connect time keeps
# streaming forever, even past its own 60-minute expiry or past an admin
# deactivating the user, since a WS connection has no per-request auth
# check the way every REST endpoint does.
REAUTH_INTERVAL_SECONDS = 60


def _validate_token(token):
    """
    Sync JWT decode + user lookup (get_user_from_token does a blocking
    psycopg2 round-trip) -- every caller below runs this via
    run_in_threadpool, never awaits it directly. Awaiting a sync DB call
    straight from an async def would block the whole event loop for its
    duration, stalling every OTHER currently-open WS connection's relay
    and keepalives for as long as this one query takes.
    """
    db = SessionLocal()
    try:
        return get_user_from_token(token, db)
    finally:
        db.close()


async def _authenticate(websocket: WebSocket):
    """First frame must be {"token": "..."}; returns (user, token), (None, None) on failure."""
    try:
        first_message = await asyncio.wait_for(
            websocket.receive_text(), timeout=AUTH_TIMEOUT_SECONDS
        )
    except (asyncio.TimeoutError, WebSocketDisconnect):
        return None, None

    try:
        token = json.loads(first_message).get("token")
    except (json.JSONDecodeError, AttributeError):
        token = None

    if not token:
        return None, None

    user = await run_in_threadpool(_validate_token, token)
    return user, token


@router.websocket("/stream")
async def stream_events(websocket: WebSocket):
    await websocket.accept()

    user, token = await _authenticate(websocket)
    if user is None:
        await websocket.close(code=4401, reason="Unauthorized")
        return

    redis_conn = get_async_redis()
    pubsub = redis_conn.pubsub()

    try:
        await pubsub.subscribe(*CHANNELS)

        last_reauth = time.monotonic()

        while True:
            message = await pubsub.get_message(
                ignore_subscribe_messages=True,
                timeout=KEEPALIVE_SECONDS,
            )

            if message is not None:
                # Already a JSON string -- publish_event() in
                # finnhub_ingestion.py json.dumps'd it before publishing,
                # so this is a straight passthrough, not a re-encode.
                await websocket.send_text(message["data"])
            else:
                await websocket.send_json({"type": "ping"})

            if time.monotonic() - last_reauth >= REAUTH_INTERVAL_SECONDS:
                last_reauth = time.monotonic()
                if await run_in_threadpool(_validate_token, token) is None:
                    await websocket.close(code=4401, reason="Session expired")
                    return

    except (WebSocketDisconnect, RedisError):
        # Client closed the tab, or Redis dropped mid-stream -- either
        # way there's nothing left to relay to. Not an error worth
        # logging loudly; the client's own reconnect logic (if any)
        # handles picking back up.
        pass

    finally:
        # Each cleanup call guarded independently -- if Redis died
        # mid-stream (the except above already caught that from
        # get_message), unsubscribe/close below would otherwise re-raise
        # the same RedisError trying to send commands over the same dead
        # connection, escaping this finally block as an unhandled
        # exception and skipping whichever cleanup call came after it.
        try:
            await pubsub.unsubscribe(*CHANNELS)
        except RedisError:
            pass

        try:
            await pubsub.close()
        except RedisError:
            pass

        try:
            await redis_conn.close()
        except RedisError:
            pass
